plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java

```diff
@@ -253,12 +251,33 @@ private Stream<DeltaLakeSplit> getSplits(
                 });
     }
 
-    private Set<Map<String, Optional<String>>> findPartitionsWithAtMostOneFile(List<AddFileEntry> addFileEntries)
+    private static Stream<AddFileEntry> filterValidDataFilesForOptimize(List<AddFileEntry> validDataFiles, long maxScannedFileSizeInBytes)
     {
-        return addFileEntries.stream().collect(groupingBy(AddFileEntry::getCanonicalPartitionValues, counting())).entrySet().stream()
-                .filter(entry -> entry.getValue() <= 1)
-                .map(Map.Entry::getKey)
-                .collect(toImmutableSet());
+        Map<Map<String, Optional<String>>, Optional<AddFileEntry>> queuedAddFileEntriesMap = new HashMap<>();
+        return validDataFiles.stream()
+                .flatMap(addFileEntry -> {
+                    Map<String, Optional<String>> canonicalPartitionValues = addFileEntry.getCanonicalPartitionValues();
+                    if (queuedAddFileEntriesMap.containsKey(canonicalPartitionValues)) {
+                        Optional<AddFileEntry> alreadyQueuedAddFileEntry = queuedAddFileEntriesMap.get(canonicalPartitionValues);
+                        if (alreadyQueuedAddFileEntry.isEmpty()) {
+                            return Stream.of(addFileEntry);
+                        }
+                        List<AddFileEntry> concatenatedAddFileEntries = ImmutableList.<AddFileEntry>builder()
+                                .add(addFileEntry)
+                                .add(alreadyQueuedAddFileEntry.get())
+                                .build();
+                        queuedAddFileEntriesMap.put(canonicalPartitionValues, Optional.empty());
+                        return concatenatedAddFileEntries.stream();
+                    }
+
+                    if (addFileEntry.getSize() < maxScannedFileSizeInBytes) {
```

Comment:
Before the "Rewrite data file filtering logic for OPTIMIZE statements" change, `findPartitionsWithAtMostOneFile` didn't take `maxScannedFileSizeInBytes` into account.

Reply:
squashed the commits into one - the rewriting commit did actually contain changes in behavior.